db/db_impl/db_impl.cc

```diff
@@ -2707,6 +2709,25 @@ ArenaWrappedDBIter* DBImpl::NewIteratorImpl(const ReadOptions& read_options,
                                             bool allow_refresh) {
   SuperVersion* sv = cfd->GetReferencedSuperVersion(this);
 
+  TEST_SYNC_POINT("DBImpl::NewIterator:1");
+  TEST_SYNC_POINT("DBImpl::NewIterator:2");
+
+  if (snapshot == kMaxSequenceNumber) {
+    // Note that the snapshot is assigned AFTER referencing the super
+    // version because otherwise a flush happening in between may compact away
+    // data for the snapshot, so the reader would see neither data that was be
+    // visible to the snapshot before compaction nor the newer data inserted
+    // afterwards.
+    // Note that the super version might not contain all the data available
+    // to this snapshot, but in that case it can see all the data in the
+    // super version, which is a valid consistent state after the user
+    // calls NewIterator().
+    snapshot = versions_->LastSequence();
```

Comment:
We could use a couple `SyncPoint`s at the end of the `LastSequence()` function to enforce the flush happens immediately after sequence number is obtained. RIght now the sync points are in the right places but it's not obvious to maintainers how they should preserve the order in the future (esp. since their names are just 1, 2, 3, 4).

Reply:
I gave it a try and found that it would make the code much more complicated. The reason is that write also calls versions_->LastSequence(). The sync-point would be complicated if we place inside. I can move the sync point to here anyway.

Reply:
The SyncPoint marker could be useful as the write/flush happen in a different thread. Up to you.